File: ListaCompras.cpp
#include <stdio.h>
#include "ListaCompras.h"

using namespace std;

int jaComprou(ListaCompras *lista, int c, int p) {
    for (int compra : lista->listaCompras[c]) {
        if (compra == p) return 1;
    }
    return 0;
}

int criaListaCompras(char *arquivoEscolhido, ListaCompras *lista) {
    Leitura compra;
    FILE *arquivo;
                    
    arquivo = fopen(arquivoEscolhido, "r");
    if (arquivo == NULL) {
        perror("Erro ao abrir o arquivo.");
        return 0;
    }
    
    fscanf(arquivo, "%*[^\n]\n");

    while(fscanf(arquivo, "%*[^,], %49[^,] , %49[^,] , %49[^\n]\n",
                  compra.cod_cliente,
                  compra.cod_produto,
                  compra.nome_produto) == 3) {

        if (lista->mapaClientes.find(compra.cod_cliente) == lista->mapaClientes.end()) {
            lista->mapaClientes[compra.cod_cliente] = lista->vetorClientes.size();
            lista->vetorClientes.push_back(compra.cod_cliente);
            lista->listaCompras.push_back(list<int>());
        }

        if (lista->mapaProdutos.find(compra.cod_produto) == lista->mapaProdutos.end()) {
            lista->mapaProdutos[compra.cod_produto] = lista->vetorProdutos.size();
            lista->vetorProdutos.push_back(compra.cod_produto);
            lista->nomeProdutos.push_back(compra.nome_produto);
        }
    }  
    fclose(arquivo);

    arquivo = fopen(arquivoEscolhido, "r");
    if (arquivo == NULL) {
        perror("Erro ao abrir o arquivo.");
        return 0;
    }

    fscanf(arquivo, "%*[^\n]\n");
    int indCliente, indProduto;

    while(fscanf(arquivo, "%*[^,], %49[^,] , %49[^,] , %49[^\n]\n",
                  compra.cod_cliente,
                  compra.cod_produto,
                  compra.nome_produto) == 3) {

        indCliente = lista->mapaClientes[compra.cod_cliente];
        indProduto = lista->mapaProdutos[compra.cod_produto];

        if(!jaComprou(lista, indCliente, indProduto)) {
            lista->listaCompras[indCliente].push_back(indProduto);
        }
    }

    fclose(arquivo);
    return 1;
}

void mostrarProdutosCliente(ListaCompras *lista, char *codigoCliente) {
    int indCliente;

    if (lista->mapaClientes.find(codigoCliente) == lista->mapaClientes.end()) {
        printf("Cliente %s nao esta registrado.", codigoCliente);
        return;
    }

    indCliente = lista->mapaClientes[codigoCliente];

    printf("\nProdutos que o cliente %s comprou:\n", codigoCliente);
    for (int produto : lista->listaCompras[indCliente]) {
        cout << "- " << lista->nomeProdutos[produto] << endl;
    }
}

void testadorExibeProdutos(ListaCompras *lista) {
    char cliente1[] = "99CL9Y01";
    char cliente2[] = "9O6OSM01";
    char cliente3[] = "99EF7201";

    mostrarProdutosCliente(lista, cliente1);
    mostrarProdutosCliente(lista, cliente2);
    mostrarProdutosCliente(lista, cliente3);
}

File: integracao_python/lista_compras.py
from dataclasses import dataclass, field
from pathlib import Path

@dataclass
class ListaCompras:
    vetor_clientes: list = field(default_factory=list)   # list[str]
    mapa_clientes:  dict = field(default_factory=dict)   # dict[str, int]

    vetor_produtos: list = field(default_factory=list)   # list[str]
    mapa_produtos:  dict = field(default_factory=dict)   # dict[str, int]

    nome_produtos:  list = field(default_factory=list)   # list[str]
    lista_compras:  list = field(default_factory=list)   # list[list[int]]


def cria_lista_compras(caminho_arquivo: str) -> ListaCompras:
    lista = ListaCompras()
    caminho = Path(caminho_arquivo)

    # Passagem 1: popula vetores e mapas de clientes e produtos
    with caminho.open("r", encoding="utf-8") as f:
        for numero_linha, linha in enumerate(f, start=1):
            linha = linha.strip()
            if not linha:
                continue

            if numero_linha == 1: # pula primeira linha (cabeçalho do csv)
                continue

            partes = linha.split(",")
            if len(partes) < 4:
                continue

            cod_cliente  = partes[1]
            cod_produto  = partes[2]
            nome_produto = partes[3]

            if cod_cliente not in lista.mapa_clientes:
                lista.mapa_clientes[cod_cliente] = len(lista.vetor_clientes)
                lista.vetor_clientes.append(cod_cliente)
                lista.lista_compras.append([])   # lista vazia para esse cliente

            if cod_produto not in lista.mapa_produtos:
                lista.mapa_produtos[cod_produto] = len(lista.vetor_produtos)
                lista.vetor_produtos.append(cod_produto)
                lista.nome_produtos.append(nome_produto)

    # Passagem 2: preenche lista_compras
    with caminho.open("r", encoding="utf-8") as f:
        for numero_linha, linha in enumerate(f, start=1):
            linha = linha.strip()
            if not linha:
                continue
            if numero_linha == 1 and linha.startswith("DATA_COMPRA"):
                continue

            partes = linha.split(",")
            if len(partes) < 4:
                continue
            
            cod_cliente  = partes[1]
            cod_produto  = partes[2]
            nome_produto = partes[3]

            ind_cliente = lista.mapa_clientes[cod_cliente]
            ind_produto  = lista.mapa_produtos[cod_produto]

            # evita duplicatas (equivalente ao jaComprou do C++)
            if ind_produto not in lista.lista_compras[ind_cliente]:
                lista.lista_compras[ind_cliente].append(ind_produto)

    return lista


def mostrar_produtos_cliente(lista: ListaCompras, cod_cliente) -> None:
    if cod_cliente not in lista.mapa_clientes:
        print(f"Cliente {cod_cliente} não está registrado.")
        return

    ind_cliente = lista.mapa_clientes[cod_cliente]
    print(f"\nProdutos que o cliente {cod_cliente} comprou:")
    for ind_produto in lista.lista_compras[ind_cliente]:
        print(f"- {lista.nome_produtos[ind_produto]}")


def testador_exibe_produtos(lista: ListaCompras) -> None:
    for cod in ["99DIQV01", "99KQAA01", "99FT8Z01"]:
        mostrar_produtos_cliente(lista, cod)